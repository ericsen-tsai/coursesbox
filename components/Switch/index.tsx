import { ChangeEvent } from 'react'
import styled from '@emotion/styled'

import { useId } from '@/hooks/useId'
import { boxShadow, transition } from '../styles'

const Wrapper = styled.label`
  & input {
    display: none;
  }
  & input:checked {
    & ~ label {
      background: ${({ theme }) => theme.components.primary};
      &::after {
        margin-left: 3.5rem;
        background: ${({ theme }) => theme.components.active};
      }
    }
  }
`

const VisiblePart = styled.label`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  cursor: pointer;
  height: 3rem;
  width: 6rem;
  border-radius: 1.6rem;
  background: ${({ theme }) => theme.components.background};
  ${({ theme }) =>
    boxShadow(theme.components.shadow1, theme.components.shadow2, false)}
  &::after {
    content: '';
    margin-left: 0.5rem;
    width: 2.1rem;
    height: 2.1rem;
    border-radius: 50%;
    background: ${({ theme }) => theme.components.nonActive};
    ${transition()}
  }
`

type Props = {
  onChange: (event: ChangeEvent<HTMLInputElement>) => void
}

function Switch({ onChange }: Props) {
  const fieldId = useId()

  return (
    <Wrapper>
      <input id={fieldId} type="checkbox" onChange={onChange} />
      <VisiblePart htmlFor={fieldId} data-testid="SwitchVisiblePart" />
    </Wrapper>
  )
}

export default Switch
