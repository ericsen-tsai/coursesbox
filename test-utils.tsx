import React, { FC, ReactElement, ReactNode } from 'react'

import { render, RenderOptions } from '@testing-library/react'

import { Themes } from './styles/themes'
import { ThemeProvider } from '@emotion/react'

const Wrapper = ({ children }: { children: ReactNode }) => {
  return <ThemeProvider theme={Themes.light}>{children}</ThemeProvider>
}

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: Wrapper, ...options })

export * from '@testing-library/react'

export { customRender as render }
