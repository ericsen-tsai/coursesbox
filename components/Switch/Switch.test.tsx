import { render, screen } from '@/test-utils'
import userEvent from '@testing-library/user-event'

import Switch from './index'

describe('Switch test cases', () => {
  it('Render check', () => {
    const onChange = jest.fn()
    jest.spyOn(Math, 'random').mockReturnValue(0.99999999999)
    const { asFragment } = render(<Switch onChange={onChange} />)

    expect(asFragment()).toMatchSnapshot()
  })

  it('Render check', async () => {
    const onChange = jest.fn()

    render(<Switch onChange={onChange} />)

    const element = screen.getByTestId('SwitchVisiblePart')

    await userEvent.click(element)
    expect(onChange).toHaveBeenCalled()
  })
})
