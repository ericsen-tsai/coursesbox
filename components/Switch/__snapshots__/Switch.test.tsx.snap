// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch test cases Render check 1`] = `
<DocumentFragment>
  <label
    class="css-4eji24"
  >
    <input
      id="fffffffff5014"
      type="checkbox"
    />
    <label
      class="css-1fjxm02"
      data-testid="SwitchVisiblePart"
      for="fffffffff5014"
    />
  </label>
</DocumentFragment>
`;
